File: python/binary-search-tree/binary_search_tree.py
"""
Binary Search Tree implementation
"""

class TreeNode:
    """
    A node in a Binary Search Tree
    """
    def __init__(self, data, left=None, right=None):
        """
        Initialize a Binary Search Tree node.
        If you pass in a left or right node, you must ensure the nodes are properly ordered or 
        you may get undesired behavior
        :param data: The value to use for this node.
        :param left: A TreeNode to use as the left side, normally None in which case left will be created on insert.
        :param right: A TreeNode to use as the right side, normally None in which case right will be created on insert.
        """
        self.data = data
        self.left = left
        self.right = right

    def insert(self, data):
        """
        Insert a value into the binary search tree using current node as root.
        :param data: The data to insert as a new node.
        """
        if data <= self.data:
            if self.left is None:
                self.left = TreeNode(data)
            else:
                self.left.insert(data)
        elif data > self.data:
            if self.right is None:
                self.right = TreeNode(data)
            else:
                self.right.insert(data)
        
    def walk(self):
        """
        Walk the tree returning all of the nodes in order
        :returns: List of nodes in sorted order.
        """
        ret = [] 
        if self.left is not None:
            ret.extend(self.left.walk())
        ret.append(self.data)
        if self.right is not None:
            ret.extend(self.right.walk())
        
        return ret

    def __str__(self):
        """
        Return this TreeNode as a string. This will implicily call the 
        __str__ functions on the left and right nodes too.
        """
        return f'TreeNode(data={self.data}, left={self.left}, right={self.right})'


class BinarySearchTree:
    """
    BinarySearchTree class - holds TreeNode objects as nodes.
    """
    def __init__(self, tree_data):
        """
        Initialize a binary search tree.
        :param tree_data: list of values to sequentially add to the tree.
        :returns: Root of binary search tree."""

        self.root = None
        for data in tree_data:
            if self.root is None:
                self.root = TreeNode(data)
            else:
                self.root.insert(data)

    def data(self):
        """
        Return the data for the root of the tree.
        """
        return self.root

    def sorted_data(self):
        """
        Return a list of data for the root node and all nodes to the left and right.
        Will be an in-order walk resulting in a sorted list for a valid binary search tree
        """
        return self.root.walk()

File: go/binary-search-tree/binary_search_tree.go
// Binary Search Tree exercise.
package binarysearchtree

// Struct for Binary Search Tree
type BinarySearchTree struct {
	left  *BinarySearchTree
	data  int
	right *BinarySearchTree
}

// NewBst creates and returns a new BinarySearchTree.
// @param i: data for the Binary Search Tree node created.
// @returns: New Binary Search Tree node with i as data and left & right = nil
func NewBst(i int) *BinarySearchTree {
	return &BinarySearchTree{left: nil, right: nil, data: i}
}

// Insert inserts an int into the BinarySearchTree.
// I am not sure why we need to allow multiple copies of the same value in the tree.
// Seems wrong to me.
// @param i: data to insert into the tree.
func (bst *BinarySearchTree) Insert(i int) {
	if bst.data >= i {
		if bst.left == nil {
			bst.left = NewBst(i)
		} else {
			bst.left.Insert(i)
		}
	} else if bst.data < i {
		if bst.right == nil {
			bst.right = NewBst(i)
		} else {
			bst.right.Insert(i)
		}
	}
}

// SortedData returns the ordered contents of BinarySearchTree as an []int.
// The values are in increasing order starting with the lowest int value.
// A BinarySearchTree that has the numbers [1,3,7,5] added will return the
// []int [1,3,5,7].
// @returns: The data from the tree node and it's children in sorted order
// as a slice of int.
func (bst *BinarySearchTree) SortedData() []int {
	var result []int = []int{}
	if bst.left != nil {
		result = append(result, bst.left.SortedData()...)
	}
	result = append(result, bst.data)
	if bst.right != nil {
		result = append(result, bst.right.SortedData()...)
	}
	return result
}
